source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,875 @@
+handshake and authentication commands.
+
+Server Monitoring
+-----------------
+
+Drivers MUST NOT use ``timeoutMS`` for commands executed by the server
+monitoring and RTT calculation threads.
+
+See `Monitoring threads do not use timeoutMS`_.
+
+Cursors
+-------
+
+For operations that create cursors, ``timeoutMS`` can either cap the lifetime
+of the cursor or be applied separately to the original operation and all
+``next`` calls. To support both of these use cases, these operations MUST
+support a ``timeoutMode`` option. This option is an enum with possible values
+``CURSOR_LIFETIME`` and ``ITERATION``. The default value depends on the type
+of cursor being created. Drivers MUST error if ``timeoutMode`` is set and
+``timeoutMS`` is not.
+
+When applying the ``timeoutMS`` option to ``next`` calls on cursors, drivers
+MUST ensure it applies to the entire call, not individual commands. For
+drivers that send ``getMore`` requests in a loop when iterating tailable
+cursors, the timeout MUST apply to the totality of all ``getMore``’s, not to
+each one individually. If a resume is required for a ``next`` call on a
+change stream, the timeout MUST apply to the entirety of the initial
+``getMore`` and all commands sent as part of the resume attempt.
+
+For ``close`` methods, drivers MUST allow ``timeoutMS`` to be overridden if
+doing so is possible in the language. If ``timeoutMS`` is explicitly
+overridden, it MUST be honored. Otherwise, if ``timeoutMS`` was applied to
+the operation that created the cursor, drivers MUST use ``min(timeoutMS,
+connectTimeoutMS)`` as the ``timeoutMS`` value for the ``killCursors``
+command if one is required. This applies to explicit ``close`` methods that
+can be called by users as well as implicit destructors that are automatically
+invoked when exiting resource blocks.
+
+See `Cursor close() methods use connectTimeoutMS as timeoutMS`_.
+
+Non-tailable Cursors
+~~~~~~~~~~~~~~~~~~~~
+
+For non-tailable cursors, the default value of ``timeoutMode`` is
+``CURSOR_LIFETIME``. If ``timeoutMS`` is set, drivers MUST apply it to the
+original operation and the lifetime of the created cursor. For example, if a
+``find`` is executed at time ``T``, the ``find`` and all ``getMore``’s on the
+cursor must finish by time ``T + timeoutMS``. When executing ``next`` calls
+on the cursor, drivers MUST use the remaining timeout as the ``timeoutMS``
+value for the operation but MUST NOT append a ``maxTimeMS`` field to
+``getMore`` commands. If there are documents remaining in a previously
+retrieved batch, the ``next`` method MUST return them even if the timeout has
+expired and MUST only return a timeout error if a ``getMore`` is required.
+
+If ``timeoutMode`` is set to ``ITERATION``, drivers MUST honor the
+``timeoutMS`` option for the initial operation. If the operation is not an
+``aggregate`` with a ``$out`` or ``$merge`` pipeline stage, drivers MUST NOT
+append a ``maxTimeMS`` field to the command sent to the server. If it is,
```

Comment:
If `timeoutMode` is `ITERATION`, a `find` command will never have `maxTimeMS` auto-applied due to an explicit `timeoutMS` option. This is because applying `maxTimeMS` would actually cause the server to cap the lifetime of the cursor, so the cursor would automatically be killed server-side after the original `timeoutMS` had elapsed, which isn't the behavior we want.

The same is true for general-purpose `aggregate` commands; we don't `timeoutMS` in this mode to avoid undesirable server behavior. However, `aggregate` commands with a $out/$merge stage will never return a non-zero cursor ID, so there's no concern about iterating the created cursor. In this case, we can safely apply `maxTimeMS` without worrying about server-side cursor behavior.

Reply:
That makes sense. I'll note the CRUD spec does say re: `$out` and `$merge`:

> Drivers MAY setup a cursor to be executed upon iteration against the output collection such that if a user were to iterate the return value, results would be returned.

Are you expecting `timeoutMS` to apply to such a cursor? If, perhaps you can change the last sentence to:

> After the operation has executed, the original `timeoutMS` value MUST be applied to each `next` call on the created cursor (including a cursor opened on the `$out`/`$merge` output collection if applicable).

I don't think we need to clarify that `maxTimeMS` shouldn't apply to the `find` query used to create that extra cursor, but feel free to reword if you want to be safe.

If you don't think this is worth addressing at all, feel free to resolve as-is.

Reply:
This is a good question. When running a $out/$merge aggregation with `timeoutMode=ITERATION`, I think the actual `aggregate` command can have a `maxTimeMS` field because it will not establish a cursor on the server. My understanding of that section of the CRUD spec is that some drivers can execute an additional `find` operation against the $out/$merge target collection. I'll bring this up in Slack offline to figure out if we should add guidance to the spec.

Reply:
Aye, I don't see any problem with specifying `maxTimeMS` for an `$out`/`$merge` pipeline. No cursor is created so we can be certain that the timeout will only apply to the pipeline's execution (similar to any other write command specifying `maxTimeMS`).

And I understand why we don't want to specify `maxTimeMS` for operations that yield cursors. So I think the question is just whether this spec should provide the guidance to remind drivers that execute an extra `find` after `$out`/`$merge` to _not_ specify `maxTimeMS` for that `find`.

Feel free to resolve this on your own once you decide whether or not to add a note.

Reply:
This thread was referring to outdated text. Are the points in my last comment addressed such that we can resolve?